server/src/main/java/io/crate/execution/ddl/tables/AlterTableClient.java

```diff
@@ -211,42 +218,69 @@ private CompletableFuture<Long> resize(BoundAlterTable analysis) {
                 String.format(Locale.ENGLISH, "Table/Partition '%s' does not exist", table.ident().fqn()));
         }
 
-        String staleIndexUUID = null;
-        for (var cursor : metadata.indices().values()) {
-            IndexMetadata indexMetadata = cursor.value;
-            Settings settings = indexMetadata.getSettings();
-            String sourceUUID = IndexMetadata.INDEX_RESIZE_SOURCE_UUID.get(settings);
-            if (sourceIndexMetadata.getIndexUUID().equals(sourceUUID)) {
-                staleIndexUUID = indexMetadata.getIndexUUID();
-                break;
+        Callable<CompletableFuture<Long>> doResize = () -> {
+            String staleIndexUUID = null;
+            for (var cursor : clusterService.state().metadata().indices().values()) {
+                IndexMetadata indexMetadata = cursor.value;
+                Settings settings = indexMetadata.getSettings();
+                String sourceUUID = IndexMetadata.INDEX_RESIZE_SOURCE_UUID.get(settings);
+                if (sourceIndexMetadata.getIndexUUID().equals(sourceUUID)) {
+                    staleIndexUUID = indexMetadata.getIndexUUID();
+                    break;
+                }
             }
-        }
+            final int targetNumberOfShards = getNumberOfShards(analysis.settings());
+            validateNumberOfShardsForResize(sourceIndexMetadata, targetNumberOfShards);
+            validateReadOnlyIndexForResize(sourceIndexMetadata);
 
-        final int targetNumberOfShards = getNumberOfShards(analysis.settings());
-        validateNumberOfShardsForResize(sourceIndexMetadata, targetNumberOfShards);
-        validateReadOnlyIndexForResize(sourceIndexMetadata);
+            final ResizeRequest request = new ResizeRequest(
+                table.ident(),
+                partitionName == null ? List.of() : partitionName.values(),
+                targetNumberOfShards
+            );
+            GenericProperties<Object> withProperties = analysis.withProperties();
+            Object timeout = withProperties.get("timeout");
+            if (timeout != null) {
+                TimeValue timeValue = TimeValue.parseTimeValue(timeout.toString(), "timeout");
+                request.timeout(timeValue);
+                request.masterNodeTimeout(timeValue);
+            }
+            CompletableFuture<ResizeResponse> resizeFuture;
+            if (staleIndexUUID == null) {
+                resizeFuture = client.execute(TransportResize.ACTION, request);
+            } else {
+                GCDanglingArtifactsRequest gcReq = new GCDanglingArtifactsRequest(List.of(staleIndexUUID));
+                resizeFuture = client.execute(TransportGCDanglingArtifacts.ACTION, gcReq)
+                    .thenCompose(_ -> client.execute(TransportResize.ACTION, request));
+            }
+            return resizeFuture.thenApply(_ -> 0L);
+        };
+
+        Consumer<Throwable> kill = _ -> {
+            String staleIndexUUID = null;
```

Comment:
Any suggestions how to test this without having some sort of long running resize?
Could make it visible on top level and create a unit test that asserts the `client.execute` call via mocks?

Reply:
Added a fixup with an integration test where I simply set the shards too high so it would run into timeouts due to not being able to allocate all shards.